File: MPChartLib/src/test/java/com/github/mikephil/charting/test/ChartDataTest.java
package com.github.mikephil.charting.test;

import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.data.ScatterData;
import com.github.mikephil.charting.data.ScatterDataSet;

import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertTrue;

/**
 * Created by philipp on 06/06/16.
 */
public class ChartDataTest {

    @Test
    public void testDynamicChartData() {

        List<Entry> entries1 = new ArrayList<Entry>();
        entries1.add(new Entry(10, 10));
        entries1.add(new Entry(15, -2));
        entries1.add(new Entry(21, 50));

        ScatterDataSet set1 = new ScatterDataSet(entries1, "");

        List<Entry> entries2 = new ArrayList<Entry>();
        entries2.add(new Entry(-1, 10));
        entries2.add(new Entry(10, 2));
        entries2.add(new Entry(20, 5));

        ScatterDataSet set2 = new ScatterDataSet(entries2, "");

        ScatterData data = new ScatterData(set1, set2);

        assertEquals(-2, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(50f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        assertEquals(6, data.getEntryCount());

        assertEquals(-1f, data.getXMin(), 0.01f);
        assertEquals(21f, data.getXMax(), 0.01f);

        assertEquals(-2f, data.getYMin(), 0.01f);
        assertEquals(50f, data.getYMax(), 0.01f);

        assertEquals(3, data.getMaxEntryCountSet().getEntryCount());


        // now add and remove values
        data.addEntry(new Entry(-10, -10), 0);

        assertEquals(set1, data.getMaxEntryCountSet());
        assertEquals(4, data.getMaxEntryCountSet().getEntryCount());

        assertEquals(-10f, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(50f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        assertEquals(-10f, data.getXMin(), 0.01f);
        assertEquals(21f, data.getXMax(), 0.01f);

        assertEquals(-10f, data.getYMin(), 0.01f);
        assertEquals(50f, data.getYMax(), 0.01f);

        data.addEntry(new Entry(-100, 100), 0);
        data.addEntry(new Entry(0, -100), 0);

        assertEquals(-100f, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(100f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        // right axis will adapt left axis values if no right axis values are present
        assertEquals(-100, data.getYMin(YAxis.AxisDependency.RIGHT), 0.01f);
        assertEquals(100f, data.getYMax(YAxis.AxisDependency.RIGHT), 0.01f);

        List<Entry> entries3 = new ArrayList<Entry>();
        entries3.add(new Entry(0, 200));
        entries3.add(new Entry(0, -50));

        ScatterDataSet set3 = new ScatterDataSet(entries3, "");
        set3.setAxisDependency(YAxis.AxisDependency.RIGHT);

        data.addDataSet(set3);

        assertEquals(3, data.getDataSetCount());

        assertEquals(-100f, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(100f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        assertEquals(-50f, data.getYMin(YAxis.AxisDependency.RIGHT), 0.01f);
        assertEquals(200f, data.getYMax(YAxis.AxisDependency.RIGHT), 0.01f);

        LineData lineData = new LineData();

        assertEquals(Float.MAX_VALUE, lineData.getYMin(), 0.01f);
        assertEquals(-Float.MAX_VALUE, lineData.getYMax(), 0.01f);

        assertEquals(Float.MAX_VALUE, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(-Float.MAX_VALUE, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        assertEquals(Float.MAX_VALUE, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f);
        assertEquals(-Float.MAX_VALUE, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f);

        assertEquals(0, lineData.getDataSetCount());

        List<Entry> lineEntries1 = new ArrayList<Entry>();
        lineEntries1.add(new Entry(10, 90));
        lineEntries1.add(new Entry(1000, 1000));

        LineDataSet lineSet1 = new LineDataSet(lineEntries1, "");

        lineData.addDataSet(lineSet1);

        assertEquals(1, lineData.getDataSetCount());
        assertEquals(2, lineSet1.getEntryCount());
        assertEquals(2, lineData.getEntryCount());

        assertEquals(10, lineData.getXMin(), 0.01f);
        assertEquals(1000f, lineData.getXMax(), 0.01f);

        assertEquals(90, lineData.getYMin(), 0.01f);
        assertEquals(1000, lineData.getYMax(), 0.01f);

        assertEquals(90, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        assertEquals(90, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f);
        assertEquals(1000, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f);

        List<Entry> lineEntries2 = new ArrayList<Entry>();
        lineEntries2.add(new Entry(-1000, 2000));
        lineEntries2.add(new Entry(2000, -3000));

        Entry e = new Entry(-1000, 2500);
        lineEntries2.add(e);

        LineDataSet lineSet2 = new LineDataSet(lineEntries2, "");
        lineSet2.setAxisDependency(YAxis.AxisDependency.RIGHT);

        lineData.addDataSet(lineSet2);

        assertEquals(2, lineData.getDataSetCount());
        assertEquals(3, lineSet2.getEntryCount());
        assertEquals(5, lineData.getEntryCount());

        assertEquals(-1000, lineData.getXMin(), 0.01f);
        assertEquals(2000, lineData.getXMax(), 0.01f);

        assertEquals(-3000, lineData.getYMin(), 0.01f);
        assertEquals(2500, lineData.getYMax(), 0.01f);

        assertEquals(90, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        assertEquals(-3000, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f);
        assertEquals(2500, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f);

        assertTrue(lineData.removeEntry(e, 1));

        assertEquals(-1000, lineData.getXMin(), 0.01f);
        assertEquals(2000, lineData.getXMax(), 0.01f);

        assertEquals(-3000, lineData.getYMin(), 0.01f);
        assertEquals(2000, lineData.getYMax(), 0.01f);

        assertEquals(90, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f);
        assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f);

        assertEquals(-3000, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f);
        assertEquals(2000, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f);
    }
}

File: MPChartLib/src/test/java/com/github/mikephil/charting/test/ChartDataTest.kt
package com.github.mikephil.charting.test

import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.data.ScatterData
import com.github.mikephil.charting.data.ScatterDataSet
import org.junit.Assert
import org.junit.Test

class ChartDataTest {
    @Test
    fun testDynamicChartData() {
        val entries1: MutableList<Entry> = ArrayList()
        entries1.add(Entry(10f, 10f))
        entries1.add(Entry(15f, -2f))
        entries1.add(Entry(21f, 50f))

        val set1 = ScatterDataSet(entries1, "")

        val entries2: MutableList<Entry> = ArrayList()
        entries2.add(Entry(-1f, 10f))
        entries2.add(Entry(10f, 2f))
        entries2.add(Entry(20f, 5f))

        val set2 = ScatterDataSet(entries2, "")

        val data = ScatterData(set1, set2)

        Assert.assertEquals(-2f, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(50f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(6, data.entryCount)

        Assert.assertEquals(-1f, data.xMin, 0.01f)
        Assert.assertEquals(21f, data.xMax, 0.01f)

        Assert.assertEquals(-2f, data.yMin, 0.01f)
        Assert.assertEquals(50f, data.yMax, 0.01f)

        Assert.assertEquals(3, data.maxEntryCountSet.entryCount)

        // now add and remove values
        data.addEntry(Entry(-10f, -10f), 0)

        Assert.assertEquals(set1, data.maxEntryCountSet)
        Assert.assertEquals(4, data.maxEntryCountSet.entryCount)

        Assert.assertEquals(-10f, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(50f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(-10f, data.xMin, 0.01f)
        Assert.assertEquals(21f, data.xMax, 0.01f)

        Assert.assertEquals(-10f, data.yMin, 0.01f)
        Assert.assertEquals(50f, data.yMax, 0.01f)

        data.addEntry(Entry(-100f, 100f), 0)
        data.addEntry(Entry(0f, -100f), 0)

        Assert.assertEquals(-100f, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(100f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        // right axis will adapt left axis values if no right axis values are present
        Assert.assertEquals(-100f, data.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(100f, data.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        val entries3: MutableList<Entry> = ArrayList()
        entries3.add(Entry(0f, 200f))
        entries3.add(Entry(0f, -50f))

        val set3 = ScatterDataSet(entries3, "")
        set3.axisDependency = YAxis.AxisDependency.RIGHT

        data.addDataSet(set3)

        Assert.assertEquals(3, data.dataSetCount)

        Assert.assertEquals(-100f, data.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(100f, data.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(-50f, data.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(200f, data.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        val lineData = LineData()

        Assert.assertEquals(Float.MAX_VALUE, lineData.yMin, 0.01f)
        Assert.assertEquals(-Float.MAX_VALUE, lineData.yMax, 0.01f)

        Assert.assertEquals(Float.MAX_VALUE, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(-Float.MAX_VALUE, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(Float.MAX_VALUE, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(-Float.MAX_VALUE, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        Assert.assertEquals(0, lineData.dataSetCount)

        val lineEntries1: MutableList<Entry> = ArrayList()
        lineEntries1.add(Entry(10f, 90f))
        lineEntries1.add(Entry(1000f, 1000f))

        val lineSet1 = LineDataSet(lineEntries1, "")

        lineData.addDataSet(lineSet1)

        Assert.assertEquals(1, lineData.dataSetCount)
        Assert.assertEquals(2, lineSet1.entryCount)
        Assert.assertEquals(2, lineData.entryCount)

        Assert.assertEquals(10f, lineData.xMin, 0.01f)
        Assert.assertEquals(1000f, lineData.xMax, 0.01f)

        Assert.assertEquals(90f, lineData.yMin, 0.01f)
        Assert.assertEquals(1000f, lineData.yMax, 0.01f)

        Assert.assertEquals(90f, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(90f, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        val lineEntries2: MutableList<Entry> = ArrayList()
        lineEntries2.add(Entry(-1000f, 2000f))
        lineEntries2.add(Entry(2000f, -3000f))

        val e = Entry(-1000f, 2500f)
        lineEntries2.add(e)

        val lineSet2 = LineDataSet(lineEntries2, "")
        lineSet2.axisDependency = YAxis.AxisDependency.RIGHT

        lineData.addDataSet(lineSet2)

        Assert.assertEquals(2, lineData.dataSetCount)
        Assert.assertEquals(3, lineSet2.entryCount)
        Assert.assertEquals(5, lineData.entryCount)

        Assert.assertEquals(-1000f, lineData.xMin, 0.01f)
        Assert.assertEquals(2000f, lineData.xMax, 0.01f)

        Assert.assertEquals(-3000f, lineData.yMin, 0.01f)
        Assert.assertEquals(2500f, lineData.yMax, 0.01f)

        Assert.assertEquals(90f, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(-3000f, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(2500f, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        Assert.assertTrue(lineData.removeEntry(e, 1))

        Assert.assertEquals(-1000f, lineData.xMin, 0.01f)
        Assert.assertEquals(2000f, lineData.xMax, 0.01f)

        Assert.assertEquals(-3000f, lineData.yMin, 0.01f)
        Assert.assertEquals(2000f, lineData.yMax, 0.01f)

        Assert.assertEquals(90f, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(-3000f, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(2000f, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        Assert.assertEquals(2, lineData.dataSetCount)
        Assert.assertTrue(lineData.removeDataSet(lineSet2))
        Assert.assertEquals(1, lineData.dataSetCount)

        Assert.assertEquals(10f, lineData.xMin, 0.01f)
        Assert.assertEquals(1000f, lineData.xMax, 0.01f)

        Assert.assertEquals(90f, lineData.yMin, 0.01f)
        Assert.assertEquals(1000f, lineData.yMax, 0.01f)

        Assert.assertEquals(90f, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(90f, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(1000f, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        Assert.assertTrue(lineData.removeDataSet(lineSet1))
        Assert.assertEquals(0, lineData.dataSetCount)

        Assert.assertEquals(Float.MAX_VALUE, lineData.xMin, 0.01f)
        Assert.assertEquals(-Float.MAX_VALUE, lineData.xMax, 0.01f)

        Assert.assertEquals(Float.MAX_VALUE, lineData.yMin, 0.01f)
        Assert.assertEquals(-Float.MAX_VALUE, lineData.yMax, 0.01f)

        Assert.assertEquals(Float.MAX_VALUE, lineData.getYMin(YAxis.AxisDependency.LEFT), 0.01f)
        Assert.assertEquals(-Float.MAX_VALUE, lineData.getYMax(YAxis.AxisDependency.LEFT), 0.01f)

        Assert.assertEquals(Float.MAX_VALUE, lineData.getYMin(YAxis.AxisDependency.RIGHT), 0.01f)
        Assert.assertEquals(-Float.MAX_VALUE, lineData.getYMax(YAxis.AxisDependency.RIGHT), 0.01f)

        Assert.assertFalse(lineData.removeDataSet(lineSet1))
        Assert.assertFalse(lineData.removeDataSet(lineSet2))
    }
}
